kolibri/plugins/coach/assets/src/views/exams-page/change-exam-visibility-modal.vue

```diff
@@ -83,45 +83,61 @@
       },
     },
     methods: {
-      visibleToClass() {
+      classInitiallySelected() {
         if (this.examVisibility.class) {
           return true;
         }
         return false;
       },
-      visibleToGroups() {
+      initiallySelectedGroups() {
         return this.examVisibility.groups.map(
             group => ({ label: group.collection.name, id: group.collection.id }));
       },
+      handleSelectChange() {
+        this.classIsSelected = !this.groupsSelected.length;
+      },
       deselectGroups() {
         this.groupsSelected = [];
       },
-      deselectClass() {
-        this.classSelected = false;
-      },
-      close() {
-        this.displayModal(false);
-      },
       updateVisibility() {
-        if (this.classSelected) {
-          const classCollection = { id: Number(this.classId), name: this.className, kind: 'classroom' };
-          this.assignExamToClass(this.examId, classCollection, this.examVisibility.groups);
+        if (this.classIsSelected) {
+          if (this.classIsSelected !== this.classInitiallySelected()) {
+            const classCollection = [{ id: this.classId, name: this.className, kind: 'classroom' }];
+            const groupAssignments = this.examVisibility.groups.map(
+              assignment => assignment.assignmentId);
+            this.updateExamAssignments(this.examId, classCollection, groupAssignments);
+          }
         } else if (this.groupsSelected.length) {
-          const groupCollections = this.groupsSelected.map(groupSelected => {
-            const collection = this.classGroups.find(
-                classGroup => classGroup.id === groupSelected.id);
-            collection.kind = 'learnergroup';
-            return collection;
-          });
-          this.assignExamToGroups(this.examId, groupCollections, this.examVisibility.class);
+          const unassignGroups = this.initiallySelectedGroups().filter(
+              initialGroup => !this.groupsSelected.find(newGroup => newGroup.id === initialGroup.id));
+          const assignGroups = this.groupsSelected.filter(
+              newGroup => !this.initiallySelectedGroups().find(
+                  initialGroup => initialGroup.id === newGroup.id));
+
+          if (unassignGroups.length || assignGroups.length) {
+            const assignGroupCollections = assignGroups.map(group => ({
+              id: group.id,
+              name: group.label,
+              kind: 'learnergroup',
+            }));
+            let unassignments = unassignGroups.map(unassignGroup => this.examVisibility.groups.find(
+              group => group.collection.id === unassignGroup.id).assignmentId);
+            if (this.examVisibility.class) {
+              unassignments = unassignments.concat(this.examVisibility.class.assignmentId);
+            }
+            this.updateExamAssignments(this.examId, assignGroupCollections, unassignments);
```

Comment:
I think it sounds ideal to close after save is completed. 

It would be confusing if the text underneath does not reflect the change (even if it is changed milliseconds afterward)